File: Experiment Processing/recommender/distance_metrics/cosine_similarity.py
import math
import numpy as np

from recommender.distance_metrics.abstract_distance_metric import AbstractDistanceMetric


class CosineSimilarity(AbstractDistanceMetric):
    """
    Cosine similarity, see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cosine.html
    for more information.
    """

    def get_name(self):
        return "Cosine Similarity"

    def calculate_distance(self, track_a, track_b):

        distance_a = 0
        distance_b = 0
        distance_a_b = 0

        track_a_features = track_a.get_metadata()
        track_b_features = track_b.get_metadata()

        for attribute in self.attribute_keys:
            attribute_a = track_a_features[attribute]
            attribute_b = track_b_features[attribute]

            distance_a += attribute_a ** 2
            distance_b += attribute_b ** 2
            distance_a_b += attribute_a * attribute_b

        try:
            return distance_a_b / (math.sqrt(distance_a) * math.sqrt(distance_b))
        except ZeroDivisionError:
            return np.NaN

File: Web Survey/resources/js/recommender/distance_metrics/CosineSimilarity.js
import AbstractDistanceMetric from "./AbstractDistanceMetric";

export default class CosineSimilarity extends AbstractDistanceMetric {

    static getMetricName() {
        return "Cosine Similarity"
    }

    calculateDistance(track, otherTrack, attributes) {
        const self = this;

        let distanceA = 0;
        let distanceB = 0;
        let distanceAB = 0;

        this.attributeKeys.forEach(attribute => {

            let attributeOwn;
            let attributeOther;

            if(attribute === "loudness") {
                attributeOwn = self.fixLoudness(attributes[track][attribute]);
                attributeOther = self.fixLoudness(attributes[otherTrack][attribute])
            } else {
                attributeOwn = attributes[track][attribute];
                attributeOther = attributes[otherTrack][attribute];
            }

            distanceA += attributeOther ** 2;
            distanceB += attributeOwn ** 2;
            distanceAB += attributeOther * attributeOwn;
        });
        let distance;

        try {
            distance = distanceAB / Math.sqrt(distanceA * distanceB)
        } catch(err){
            console.log("Zero division");
            return NaN;
        }
        return distance;
    }

}